# **SET-UP AND DATA IMPORT**

In [18]:
import pandas as pd
import numpy as np
import re # module for pattern matching and text manipulation
pd.options.display.max_colwidth = 4000
pd.options.display.max_rows = 100

from sklearn.feature_extraction.text import CountVectorizer # converts text into numbers by counting how often each word (or word pair) appears in documents.
from sklearn.feature_selection import chi2 # chi-squared statistical test to measure how strongly each feature (e.g., word or n-gram) is associated with a class label, a ranking procedure to distinguish most discriminative features.

In [ ]:
# mount Google drive

Mounted at /content/drive


In [ ]:
# path = path to your Google drive

text_access = "text"

# Read only the specified text column
df_discharge_text = pd.read_csv(path, usecols=[text_access], compression='gzip')

In [4]:
# Keep only 100K records for computational efficiency.
df_discharge_text = df_discharge_text.iloc[:100000].reset_index(drop=True)
df_discharge_text.head(2)

text
0   \nName:  ___                     Unit No:   ___\n \nAdmission Date:  ___              Discharge Date:   ___\n \nDate of Birth:  ___             Sex:   F\n \nService: MEDICINE\n \nAllergies: \nNo Known Allergies / Adverse Drug Reactions\n \nAttending: ___\n \nChief Complaint:\nWorsening ABD distension and pain \n \nMajor Surgical or Invasive Procedure:\nParacentesis\n\n \nHistory of Present Illness:\n___ HCV cirrhosis c/b ascites, hiv on ART, h/o IVDU, COPD, \nbioplar, PTSD, presented from OSH ED with worsening abd \ndistension over past week.  \nPt reports self-discontinuing lasix and spirnolactone ___ weeks \nago, because she feels like "they don't do anything" and that \nshe "doesn't want to put more chemicals in her." She does not \nfollow Na-restricted diets. In the past week, she notes that she \nhas been having worsening abd distension and discomfort. She \ndenies ___ edema, or SOB, or orthopnea. She denies f/c/n/v, d/c, \ndysuria. She had food poisoning a week ago from eating stale \ncake (n/v 20 min after food ingestion), which resolved the same \nday. She denies other recent illness or sick contacts. She notes \nthat she has been noticing gum bleeding while brushing her teeth \nin recent weeks. she denies easy bruising, melena, BRBPR, \nhemetesis, hemoptysis, or hematuria.  \nBecause of her abd pain, she went to OSH ED and was transferred \nto ___ for further care. Per ED report, pt has brief period of \nconfusion - she did not recall the ultrasound or bloodwork at \nosh. She denies recent drug use or alcohol use. She denies \nfeeling confused, but reports that she is forgetful at times.  \nIn the ED, initial vitals were 98.4 70 106/63 16 97%RA  \nLabs notable for ALT/AST/AP ___ ___: ___, \nTbili1.6, WBC 5K, platelet 77, INR 1.6  \n\n \nPast Medical History:\n1. HCV Cirrhosis  \n2. No history of abnormal Pap smears.  \n3. She had calcification in her breast, which was removed  \npreviously and per patient not, it was benign.  \n4. For HIV disease, she is being followed by Dr. ___ Dr.  \n___.  \n5. COPD  \n6. Past history of smoking.  \n7. She also had a skin lesion, which was biopsied and showed  \nskin cancer per patient report and is scheduled for a complete  \nremoval of the skin lesion in ___ of this year.  \n8. She also had another lesion in her forehead with purple  \ndiscoloration. It was biopsied to exclude the possibility of  \n___'s sarcoma, the results is pending.  \n9. A 15 mm hypoechoic lesion on her ultrasound on ___  \nand is being monitored by an MRI.  \n10. History of dysplasia of anus in ___.  \n11. Bipolar affective disorder, currently manic, mild, and PTSD. \n \n12. History of cocaine and heroin use.  \n\n \nSocial History:\n___\nFamily History:\nShe a total of five siblings, but she is not  \ntalking to most of them. She only has one brother that she is in \n \ntouch with and lives in ___. She is not aware of any  \nknown GI or liver disease in her family.  \nHer last alcohol consumption was one drink two months ago. No  \nregular alcohol consumption. Last drug use ___ years ago. She  \nquit smoking a couple of years ago.  \n\n \nPhysical Exam:\nVS: 98.1 107/61 78 18 97RA  \nGeneral: in NAD  \nHEENT: CTAB, anicteric sclera, OP clear  \nNeck: supple, no LAD  \nCV: RRR,S1S2, no m/r/g  \nLungs: CTAb, prolonged expiratory phase, no w/r/r  \nAbdomen: distended, mild diffuse tenderness, +flank dullness, \ncannot percuss liver/spleen edge ___ distension  \nGU: no foley  \nExt: wwp, no c/e/e, + clubbing  \nNeuro: AAO3, converse normally, able to recall 3 times after 5 \nminutes, CN II-XII intact  \n\nDischarge:\n\nPHYSICAL EXAMINATION:  \nVS: 98 105/70 95\nGeneral: in NAD  \nHEENT: anicteric sclera, OP clear  \nNeck: supple, no LAD  \nCV: RRR,S1S2, no m/r/g  \nLungs: CTAb, prolonged expiratory phase, no w/r/r  \nAbdomen: distended but improved, TTP in RUQ, \nGU: no foley  \nExt: wwp, no c/e/e, + clubbing  \nNeuro: AAO3,  CN II-XII intact  \n\n \nPertinent Results:\n___ 10:25PM   GLUCOSE-109* UREA

In [5]:
# Create an empty data frame for results.

df_extract = df_discharge_text.drop("text", axis = 1)
df_extract

Empty DataFrame
Columns: []
Index: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, ...]

[100000 rows x 0 columns]

# **TEXT EXTRACTION AND CLEAN-UP**

From the full record, extract "Chief Complaint" and add it as a new column.

In [6]:
# User Defined Function to extract Chief Complaint: (everything after "Chief Complaint:" until the end of the line) and add it as a column to the original dataframe.

def extract_complaint(pattern, df_discharge_text, column_name="text"):
    """
    Extracts matches based on a pre-defined regex pattern from a specified column in the dataframe.

    Parameters:
        pattern (str): The regular expression pattern to use for extraction: r'Chief Complaint:\s*(.*)'
        df (pd.DataFrame): The DataFrame containing the text column: df_discharge_text.
        column_name (str): The name of the column to search, "text_access".

    Returns:
        pd.Series: A Series containing the extracted matches which will be added as a new column to the original dataframe.
    """
    # applying the pre-defined regex to each row of the target column
    return df_discharge_text[column_name].apply(lambda text: re.search(pattern, str(text)).group(1) if re.search(pattern, str(text)) else None)

# df[column_name] → Selects the target column ("text_access" in df_discharge_text).
# .apply(lambda text: ...) → Applies a function to each row in that column.
# str(text) → Converts the value to a string (in case it contains NaN or numbers).
# re.search(pattern, str(text)) → Searches for the regex pattern inside each row’s text:
# If no match is found, re.search() returns None.
# If a match is found, .group(1) extracts the first capturing group from the regex.
# if re.search(pattern, str(text)) else None → Ensures that if no match is found, it returns None instead of causing an error.

# Define regex pattern
pattern = r'Chief Complaint:\s*(.*)'

# Apply function to extract chief complaints
df_extract["Chief Complaint"] = extract_complaint(pattern, df_discharge_text, "text")

# Display results
df_extract

Chief Complaint
0      Worsening ABD distension and pain 
1       abdominal fullness and discomfort
2                   altered mental status
3                          Abdominal pain
4                   Visual hallucinations
...                                   ...
99995                Chest pain/tightness
99996          intractable L facial pain 
99997                     chest pain, SOB
99998                        leg swelling
99999            abdominal pain, diarrhea

[100000 rows x 1 columns]

From the full record, extract "Diagnosis" and add it as a new column.

In [7]:
# User Defined Function to extract Diagnoses: (multi-line after "Discharge Diagnoses" until break or Discharge Condition:) and add it as a column to the original dataframe.

def extract_all_dz(pattern, df_discharge_text, column_name="text"):
    """
    Extracts matches based on a pre-defined regex pattern from a specified column in the dataframe.

    Parameters:
        pattern (str): The regular expression pattern to use for extraction:
        df (pd.DataFrame): The DataFrame containing the text column: df_discharge_text.
        column_name (str): The name of the column to search, "text_access".

    Returns:
        pd.Series: A Series containing the extracted matches which will be added as a new column to the original dataframe.
    """
    # applying the pre-defined regex to each row of the target column
    return df_discharge_text[column_name].apply(lambda text: re.search(pattern, str(text)).group(1) if re.search(pattern, str(text)) else None)


# Define regex pattern
pattern = pattern = r'(?s)Discharge Diagnosis:\s*(.+?)(Discharge Condition:|\Z)'

# \n\s*\n stops at the first blank line (double newline) to prevent capturing unwanted sections.
# \Z ensures it captures everything until the end if no blank line is found.

# Apply function to extract diagnosis
df_extract["All Dz"] = extract_all_dz(pattern, df_discharge_text, "text")

# Display results
df_extract

Chief Complaint  \
0      Worsening ABD distension and pain    
1       abdominal fullness and discomfort   
2                   altered mental status   
3                          Abdominal pain   
4                   Visual hallucinations   
...                                   ...   
99995                Chest pain/tightness   
99996          intractable L facial pain    
99997                     chest pain, SOB   
99998                        leg swelling   
99999            abdominal pain, diarrhea   

                                                                                                                                                                                                                                                                 All Dz  
0                                                                                                                                                                                                                                        Ascites from Portal HTN\n\n \n  
1                                                                                                                                                                       Primary:  diuretic refractory ascites\nSecondary: HCV cirrhosis, HIV, hyponatremia, COPD\n\n \n  
2                                                                                                                                                                     Primary:\nHypotension\nHyperkalemia\nAcute Kidney Injury\n\nSecondary:\nHIV\nCirrhosis\nCOPD\n \n  
3                                                                                                      PRIMARY DIAGNOSIS\n=================\n-Decompensated HCV cirrhosis\n-Hyperkalemia\n\nSECONDARY DIAGNOSES\n===================\n-HIV\n-COPD\n-Hyponatremia\n\n \n  
4                                                                                                                                                                                                                                                     ___ Dementia\n \n  
...                                                                                                                                                                                                                                                                 ...  
99995                      Coronary artery disease s/p Coronary artery bypass graft\nAtrial Fibrillation\nAnemia acute blood loss\n\nPast medical history:\nHypertension\nHyperlipidemia\nHypothyroidism\nDiabetes\nRemote afib history never on anticoagulation\n\n \n  
99996                                                                                                                                                                                                                    Refractory intractable left facial pain\n\n \n  
99997                                                                                                                                                               PRIMARY:\n- NSTEMI\n- Severe, symptomatic aortic stenosis\nSECONDARY:\n- Severe malnutrition\n\n \n  
99998  PRIMARY DIAGNOSIS: \n==================\nAcute on chronic HFpEF (EF 45%) \nCritical aortic stenosis (valve area 0.5, peak velocity 4.0) \nNSTEMI s/p PCI x3 (LAD, mLAD, ___ \n\nSECONDARY DIAGNOSES: \n====================\nBenign prostatic hyperplasia\n\n \n  
99999                                                                                               #Primary:\nPortal vein thrombosis\nSMV thrombosis\n\n#Secondary:\nHIT\nC.diff infection\nPneumothorax\nAcute blood loss anemia  \nAlcohol use\nTransaminitis\n\n \n  

[100000 rows x 2 columns]

In [8]:
# Since we captured all non-white space following Discharge Diagnosis, we need to remove unnecessary characters and words.
# First, find and remove special characters, to include creative patterns like "====" or "---" and digits

df_extract["All Dx"] = df_extract["All Dz"].str.replace(r'(\s*\n\s*|- |-|={3,}|-{3,}|_{3,}|#|\d+)', ' ', regex=True).str.strip()


# Second, remove "Primary diagnosis:" and "Secondary diagnosis:"
pattern = r'(?i)Primary diagnosis:|Secondary diagnosis:|Primary diagnoses:|Primary Diagnosis|Secondary Diagnosis|Secondary diagnoses:|PRIMARY DIAGNOSIS|SECONDARY DIAGNOSIS|PRIMARY DIAGNOSES|SECONDARY DIAGNOSES'

df_extract["All Dx"] = df_extract["All Dx"].str.replace(pattern, '', regex=True).str.strip()
df_extract

Chief Complaint  \
0      Worsening ABD distension and pain    
1       abdominal fullness and discomfort   
2                   altered mental status   
3                          Abdominal pain   
4                   Visual hallucinations   
...                                   ...   
99995                Chest pain/tightness   
99996          intractable L facial pain    
99997                     chest pain, SOB   
99998                        leg swelling   
99999            abdominal pain, diarrhea   

                                                                                                                                                                                                                                                                 All Dz  \
0                                                                                                                                                                                                                                        Ascites from Portal HTN\n\n \n   
1                                                                                                                                                                       Primary:  diuretic refractory ascites\nSecondary: HCV cirrhosis, HIV, hyponatremia, COPD\n\n \n   
2                                                                                                                                                                     Primary:\nHypotension\nHyperkalemia\nAcute Kidney Injury\n\nSecondary:\nHIV\nCirrhosis\nCOPD\n \n   
3                                                                                                      PRIMARY DIAGNOSIS\n=================\n-Decompensated HCV cirrhosis\n-Hyperkalemia\n\nSECONDARY DIAGNOSES\n===================\n-HIV\n-COPD\n-Hyponatremia\n\n \n   
4                                                                                                                                                                                                                                                     ___ Dementia\n \n   
...                                                                                                                                                                                                                                                                 ...   
99995                      Coronary artery disease s/p Coronary artery bypass graft\nAtrial Fibrillation\nAnemia acute blood loss\n\nPast medical history:\nHypertension\nHyperlipidemia\nHypothyroidism\nDiabetes\nRemote afib history never on anticoagulation\n\n \n   
99996                                                                                                                                                                                                                    Refractory intractable left facial pain\n\n \n   
99997                                                                                                                                                               PRIMARY:\n- NSTEMI\n- Severe, symptomatic aortic stenosis\nSECONDARY:\n- Severe malnutrition\n\n \n   
99998  PRIMARY DIAGNOSIS: \n==================\nAcute on chronic HFpEF (EF 45%) \nCritical aortic stenosis (valve area 0.5, peak velocity 4.0) \nNSTEMI s/p PCI x3 (LAD, mLAD, ___ \n\nSECONDARY DIAGNOSES: \n====================\nBenign prostatic hyperplasia\n\n \n   
99999                                                                                               #Primary:\nPortal vein thrombosis\nSMV thrombosis\n\n#Secondary:\nHIT\nC.diff infection\nPneumothorax\nAcute blood loss anemia  \nAlcohol use\nTransaminitis\n\n \n   

                                                                                                                                                                                                                            All Dx  
0                                                  

From the full record, extract the "History of Present Illness" and add it as a new column.

In [9]:
# User Defined Function to extract History of Present Illness: (everything after "History of Present Illness:" until break) and add it as a column to the original dataframe.

def extract_present_illness(pattern, df_discharge_text, column_name="text"):
    """
    Extracts matches based on a pre-defined regex pattern from a specified column in the dataframe.

    Parameters:
        pattern (str): The regular expression pattern to use for extraction: r'History of Present Illness:\s*(.*)'
        df (pd.DataFrame): The DataFrame containing the text column: df_discharge_text.
        column_name (str): The name of the column to search, "text".

    Returns:
        pd.Series: A Series containing the extracted matches which will be added as a new column to the original dataframe.
    """
    # applying the pre-defined regex to each row of the target column
    return df_discharge_text[column_name].apply(lambda text: re.search(pattern, str(text)).group(1) if re.search(pattern, str(text)) else None)


# Define regex pattern
pattern = pattern = r'(?s)History of Present Illness:\s*(.+?)(Past Medical History:|\Z)'

# \n\s*\n stops at the first blank line (double newline) to prevent capturing unwanted sections.
# \Z ensures it captures everything until the end if no blank line is found.

# Apply function to extract history of present illness
df_extract["Present Illness History Extracted"] = extract_present_illness(pattern, df_discharge_text, "text")

In [10]:
# Since we captured all text as is for History of Present Illness, we need to remove unnecessary characters.
# Find and remove special characters and new lines.

df_extract["Present Illness History"] = df_extract["Present Illness History Extracted"].str.replace(r'(\s*\n\s*|_{3,}|={3,}|\+)', ' ', regex=True).str.strip()
df_extract.iloc[4]

Chief Complaint                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     Visual hallucinations
All Dz                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                

In [11]:
df_extract.drop(["Present Illness History Extracted", "All Dz"], axis = 1, inplace = True)
df_extract.head(2)

Chief Complaint  \
0  Worsening ABD distension and pain    
1   abdominal fullness and discomfort   

                                                                                    All Dx  \
0                                                                  Ascites from Portal HTN   
1  Primary:  diuretic refractory ascites Secondary: HCV cirrhosis, HIV, hyponatremia, COPD   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      Present Illness History  
0                                                                                        HCV cirrhosis c/b ascites, hiv on ART, h/o IVDU, COPD, bioplar, PTSD, presented from OSH ED with worsening abd distension over past week. Pt reports self-discontinuing lasix and spirnolactone   weeks ago, because she feels like "they don't do anything" and that she "doesn't want to put more chemicals in her." She does not follow Na-restricted diets. In the past week, she notes that she has been having worsening abd distension and discomfort. She denies   edema, or SOB, or orthopnea. She denies f/c/n/v, d/c, dysuria. She had food poisoning a week ago from eating stale cake (n/v 20 min after food ingestion), which resolved the same day. She denies other recent illness or sick contacts. She notes that she has been noticing gum bleeding while brushing her teeth in recent weeks. she denies easy bruising, melena, BRBPR, hemetesis, hemoptysis, or hematuria. Because of her abd pain, she went to OSH ED and was transferred to   for further care. Per ED report, pt has brief period of confusion - she did not recall the ultrasound or bloodwork at osh. She denies recent drug use or alcohol use. She denies feeling confused, but reports that she is forgetful at times. In the ED, initial vitals were 98.4 70 106/63 16 97%RA Labs notable for ALT/AST/AP    :  , Tbili1.6, WBC 5K, platelet 77, INR 1.6  
1  with HIV on HAART, COPD, HCV cirrhosis complicated by ascites and HE admitted with abdominal distention and pain. She was admitted to   for the same symptoms recently and had 3L fluid removed (no SBP) three days ago and felt better. Since discharge, her abdomen has become increasingly distended with pain. This feels similar to prior episodes of ascites. Her diuretics were recently decreased on   due to worsening hyponatremia 128 and hyperkalemia 5.1. Patient states she has been compliant with her HIV and diuretic medications but never filled out the lactulose prescription. She states she has had   BMs daily at home. She has had some visual hallucinations and forgetfulness. Her appetite has been poor. In the ED, initial vitals were 98.9 88 116/88 18 97% RA. CBC near baseline, INR 1.4, Na 125, Cr 0.6. AST and ALT mildl

# **DISCRIMINATIVE FEATURE SELECTION**

Discriminative Feature Selection is a process of identifying the words, phrases, or features that best differentiate between two or more classes. In our case, the positive class is notes with chest pain complaints and the negative class is notes without such complaints. We will focus on the "Present History Illness" part of the data set to extract features differentiating the two classes.

The first method we will test is the Chi Square model which measures the strength of the relationship between a word (or n-gram) and the class label. Using class labels (1=positive class, 0 = negative class, in a binary prediction), it scores and ranks each feature (word/n-gram) on how strongly it is associated with either label.

Step 1: Review all chest-related complaints (to familiarize yourself with options).
Step 2: Extract all notes with the word "chest" in "Chief Complaint".
Step 3: Create a binary class column, Chest Complaint = 1, No Chest Complaint = 0 (qualification = presence of feature "chest").
Step 4: Run the feature discriminative selection using the Chi-square test.

In [12]:
# first, we need to identify all chest-related complaints to familiarize ourselves with choices.
df_chest = df_extract[df_extract["Chief Complaint"].fillna("").str.contains(r"\bchest\b", case=False, regex=True)]
df_chest["Chief Complaint"].unique() # cannot see the full array

complaints = df_chest["Chief Complaint"].unique().tolist()
for c in complaints:
    print(c)

Dyspnea and chest tightness
Atypical chest pain
Chest Pain  
Nonexertional Chest pain
Chest pain
Chest pain  
Substernal chest pain and throat tightness with exertion
chest pain
Chest Pain
chest pressure
Chest pain and dyspnea with exertion/Fatigue
pleuritic chest pain 
Left chest pain
DOE and Chest pressure
chest pain, fevers, chills
Exterional chest pain
Chest pain, subtherapeutic INR
shortness of breath, chest pain
Substernal Chest pain; ST Elevation myocardial infarction 
chest pain 
chest pain, dyspnea on exertion
Right sided chest pain and worsening SOB
Chills, poor po intake, hyperglycemia, chest pain, heavy 
Chest pain, DOE
chest discomfort, DOE
chest pain, N/V
Chest pain in PACU after ERCP
chest tightness and left arm tingling
chest pain dyspnea 
Dyspnea, chest pain
diarrhea, chest pain, afib with rapid ventricular response
Chest pain and dyspnea 
Chief compliant: chest pain  
Chest pain and shortness of breath x4 days
chest pain, hematochezia 
Worsening R sided chest pain x 6

In [12]:
# check records with no chief complaint
df_blank = df_extract[df_extract["Chief Complaint"] == ""]

In [13]:
# Check for blanks and na in the target column.
(df_extract["Chief Complaint"] == " ").sum()
(df_extract["Chief Complaint"].isna() | (df_extract["Chief Complaint"].str.strip() == "")).sum()

# We can drop these blank records as they constitute less than 5% of the total.
df_extract.dropna(subset=["Chief Complaint"], inplace=True)
df_extract = df_extract[df_extract["Chief Complaint"].str.strip() != ""]

In [14]:
# Doube-check blanks and na records.
(df_extract["Chief Complaint"].isna() | (df_extract["Chief Complaint"].str.strip() == "")).sum()

np.int64(0)

In [15]:
# Now we will add the label column coded as a Boolean column, based on the presence of the word "chest".

# Define regex pattern for chest-related complaints. We will use only the word "chest" which appears to be the key word paired with secondary words like pain and pressure.
pattern = r"chest"
# Option 1: any non-string records will be assigned False or 0 (No Chest Pain)
# Apply case-insensitive search using the pre-defined regex pattern to the target column.
df_extract["Chest Pain Label"] = df_extract["Chief Complaint"] \
    .fillna("").str.lower().str.contains(pattern, regex=True).astype(int)

# Option 2:
# Create a full binary column with a lambda function to explicitly assign 0 when the pattern is not found
# (including blanks or NaN). Lambda functions are a great choice here as they are quick, local methods to iterate
# over elements with functions like apply() and map().

df_extract["Chest Pain Label"] = df_extract["Chief Complaint"].apply(
    lambda x: 1 if isinstance(x, str) and re.search(pattern, x, flags=re.IGNORECASE) else 0
)

df_chest_chi = df_extract[df_extract["Chest Pain Label"] ==  1]
df_chest_chi.head()

Chief Complaint  \
20  Dyspnea and chest tightness   
46          Atypical chest pain   
47                 Chest Pain     
52     Nonexertional Chest pain   
63                   Chest pain   

                                                                                                                                                                                                                                                                                                                                                                                 All Dx  \
20                                                                                                                                                                                                                                                                                                                                Acute on chronic decompensated systolic heart failure   
46                                                                                                                                                                                                                                                                                                                                                              Coronary Artery Disease   
47                                                                                                                                                                                                                                                                                               PRIMARY:  . atypical chest pain Secondary:  . Hypertension  . Asthma  . Diverticulitis   
52                                                                                                                                                                                                                                                                                                                                               Rapid Atrial Fibrillation/Flutter COPD   
63  Chest wall pain atypical for angina  Musculoskeletal pain  Known native coronary artery and bypass graft disease  Type   Diabetes mellitus, with  Chronic kidney disease, stage    Acute kidney injury  Chronic obstructive pulmonary disease  Hypertension  Hypotension  Hypokalemia  Chronic shoulder pain on narcotics  Obstructive sleep apnea  Gastroseophageal reflux disease   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          

In [16]:
# Use the Present Illness column as input text
target_text = df_extract["Present Illness History"].fillna("").tolist()

labels = df_extract["Chest Pain Label"]

# Vectorize using unigrams and bigrams, remove English stopwords
vectorizer = CountVectorizer(ngram_range=(1, 3), stop_words='english')
X = vectorizer.fit_transform(target_text)

# Compute chi-square scores
chi2_scores, _ = chi2(X, labels)
feature_names = vectorizer.get_feature_names_out()

# Get top 20 discriminative features
top_features = sorted(zip(chi2_scores, feature_names), reverse=True)[:20]

# Display in table
pd.DataFrame(top_features, columns=["Score", "Feature"])


Score           Feature
0   50794.181044             chest
1   30318.951973        chest pain
2   14850.198115             nitro
3   13059.767085           cardiac
4   13025.476853        substernal
5   13006.469614                sl
6   12825.871626              pain
7   12138.439711               ekg
8   12071.799842       diaphoresis
9    9917.452934              cath
10   8976.729591     nitroglycerin
11   8162.486545  substernal chest
12   8128.221228               ntg
13   8010.518251               lad
14   7909.860699        exertional
15   7844.970647    chest pressure
16   7616.475683         pain free
17   7462.898496           heparin
18   7136.923672                st
19   7109.350642               asa

The Chi-square score helps with separating high-signal features from high-noise features. The higher the score, the greater is the discriminating power of a given word or, in this case, a phrase of two words. For example, the feature "cath" with the score of 9,917 is strongly associated with the positive class, "Chest Complaint", i.e., it appears with a greater frequency in discharge notes with some type of chest complaint. The score is the measure of the difference between expected and observed frequences.

In [19]:
# Compare to a method of limiting extracted and ranked features to only bi-grams, i.e., two-word phrases.
# This method gives additional insights because the word "chest" is usually paired with another word like "pressure", "pain", "discomfort".


target_text = df_extract["Present Illness History"].fillna("").tolist()
labels = df_extract["Chest Pain Label"]

# Vectorize using unigrams, bigrams, and trigrams
vectorizer = CountVectorizer(ngram_range=(1, 2), stop_words='english')
X = vectorizer.fit_transform(target_text)
feature_names = vectorizer.get_feature_names_out()

# Compute chi-square scores
chi2_scores, _ = chi2(X, labels)

# Combine scores and feature names
scored_features = list(zip(chi2_scores, feature_names))

# Filter for only bigrams (i.e., features with exactly one space)
bigram_features = [(score, feat) for score, feat in scored_features if feat.count(" ") == 1]

# Get top 20 bigrams by score
top_bigrams = sorted(bigram_features, reverse=True)[:100]

# Display
pd.DataFrame(top_bigrams, columns=["Score", "Feature"])


Score                  Feature
0   30318.951973               chest pain
1    8162.486545         substernal chest
2    7844.970647           chest pressure
3    7616.475683                pain free
4    7063.003483           presents chest
5    6363.109809              sided chest
6    6117.312665              stress test
7    5647.283292         chest discomfort
8    4896.733612  cardiac catheterization
9    4760.349997                 sl nitro
10   4434.547420           cardiac review
11   4325.777192          orthopnea ankle
12   4293.065905              ankle edema
13   4290.532831       syncope presyncope
14   4289.291885       edema palpitations
15   4287.458369     paroxysmal nocturnal
16   4272.459181        nocturnal dyspnea
17   4241.152669        dyspnea orthopnea
18   4235.266065     palpitations syncope
19   4157.669462               red stools
20   4146.768220          systems notable
21   4103.062853         hemoptysis black
22   4100.294711               stools red
23   4061.043491     thrombosis pulmonary
24   4017.820715       pulmonary embolism
25   3984.983906        embolism bleeding
26   3951.998474       exertional buttock
27   3927.903039        denies exertional
28   3917.939345             buttock calf
29   3910.633477            bleeding time
30   3845.902167             time surgery
31   3797.251200                 tia deep
32   3784.433126              deep venous
33   3772.746034         surgery myalgias
34   3763.181579           myalgias joint
35   3762.664685               stroke tia
36   3696.606465              pains cough
37   3675.614319        venous thrombosis
38   3664.743814           history stroke
39   3639.707754                calf pain
40   3597.266351              heparin gtt
41   3535.584917                 cath lab
42   3526.113049             denies prior
43   3506.131691         negative cardiac
44   3472.267228                   sl ntg
45   3451.321117          notable absence
46   3362.001222          absence dyspnea
47   3330.156647            prior history
48   3308.240409            stools denies
49   3306.239509            rigors denies
50   3275.280024          chest tightness
51   3238.586375      exertion paroxysmal
52   3182.107459         exertional chest
53   3161.855568          presented chest
54   3144.577935         presenting chest
55   3134.547248         cough hemoptysis
56   3027.434329     associated shortness
57   3014.170398             black stools
58   2998.106042           review systems
59   2934.872718           associated sob
60   2795.108768              joint pains
61   2792.720588          coronary artery
62   2778.275767            recent fevers
63   2770.877491            chills rigors
64   2710.728164          developed chest
65   2681.641906               ekg showed
66   2672.601029              onset chest
67   2636.071406                given asa
68   2599.772431           systems denies
69   2597.008699              pain review
70   2493.098624           episodes chest
71   2490.379387       intermittent chest
72   2451.077850          started heparin
73   2377.824514            st elevations
74   2318.710563               left chest
75   2313.343837             cardiac cath
76   2303.084905            reports chest
77   2277.439228       nausea diaphoresis
78   2257.239572   associated diaphoresis
79   2227.652297                   ra ekg
80   2106.871126            pain resolved
81   2044.739552            episode chest
82   2013.930538           artery disease
83   1965.455109                nitro gtt
84   1960.942761       diaphoresis nausea
85   1951.869881       absence paroxysmal
86   1949.620004                 left arm
87   1911.267968          sob diaphoresis
88   1863.424164           st depressions
89   1826.227873              ekg changes
90   1805.313514            given aspirin
91   1791.082514          unstable angina
92   1783.713410                 ed chest
93   1742.385293         sl nitroglycerin
94   1726.763008     